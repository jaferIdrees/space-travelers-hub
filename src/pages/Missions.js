import React from 'react';
import MissionsList from '../components/Missions';

const Rockets = () => (
  <MissionsList />
);

export default Rockets;
