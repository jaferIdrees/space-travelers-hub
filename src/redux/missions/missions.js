import GetMissions from '../../API/missionsService';

// Actions
const RETRIEVE_MISSIONS = 'RETRIEVE_MISSION';
const ERROR = 'ERROR';

const initialState = [];

// Reducer
export default function reducer(state = initialState, action = {}) {
  const { type, payload } = action;
  switch (type) {
    case RETRIEVE_MISSIONS:
    {
      const newState = [];
      payload.forEach((mission) => {
        const {
          // eslint-disable-next-line camelcase
          mission_id, mission_name, description,
        } = mission;
        newState.push({
          // eslint-disable-next-line camelcase
          mission_id, mission_name, description,
        });
      });
      return newState;
    }
    default: return state;
  }
}

export const retrieveMissions = () => async (dispatch) => {
  try {
    const response = await GetMissions();
    dispatch({
      type: RETRIEVE_MISSIONS,
      payload: response.data,
    });
  } catch (err) {
    dispatch({
      type: ERROR,
      payload: { err },
    });
  }
};
