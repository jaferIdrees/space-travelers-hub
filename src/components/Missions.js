import React from 'react';
import Table from 'react-bootstrap/Table';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { v4 as uuidv4 } from 'uuid';
import { useSelector, useDispatch } from 'react-redux';
import { retrieveMissions, joinMission } from '../redux/missions/missions';

function Missions() {
  const missions = useSelector((state) => state.missionsReducer);
  const dispatch = useDispatch();
  if (missions.length === 0) {
    dispatch(retrieveMissions());
  }
  return (
    <Table striped bordered hover>
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th> </th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <tr key={uuidv4()}>
            <td>{mission.mission_name}</td>
            <td>{mission.description}</td>
            <td>
              {mission.joined && (<Badge bg="primary">Active Member</Badge>)}
              {!mission.joined && (<Badge bg="secondary">Not a Member</Badge>)}
            </td>
            <td>
              <Button className={mission.joined ? 'activeButton' : 'notActiveButton'} type="button" onClick={() => dispatch(joinMission({ mission }))}>
                {mission.joined ? 'Leave Mission' : 'Join Mission'}
              </Button>

            </td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
}

export default Missions;
