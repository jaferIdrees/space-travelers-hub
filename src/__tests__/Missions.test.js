import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/configureStore';
import Missions from '../components/Missions';
import { retrieveMissions } from '../redux/missions/missions';

describe('Test missions reducers and functions', () => {
  it('Fetches data from API', async () => {
    await store.dispatch(retrieveMissions());
    expect(store.getState().rockets).toBeTruthy();
  });
});

describe('Missions component testing using snapshots', () => {
  it('Test Missions component renders', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      );
    expect(tree).toMatchSnapshot();
  });
});
