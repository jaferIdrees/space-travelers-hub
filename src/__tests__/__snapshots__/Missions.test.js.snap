// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Misi component testing using snapshots Test Rockets component renders 1`] = `
Array [
  <link
    crossOrigin="anonymous"
    href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css"
    integrity="sha384-0evHe/X+R7YkIZDRvuzKMRqM+OrBnVFBL6DOitfPri4tjfHxaWutUpFmBp4vmVor"
    rel="stylesheet"
  />,
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>,
]
`;

exports[`Missions component testing using snapshots Test Missions component renders 1`] = `
Array [
  <link
    crossOrigin="anonymous"
    href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css"
    integrity="sha384-0evHe/X+R7YkIZDRvuzKMRqM+OrBnVFBL6DOitfPri4tjfHxaWutUpFmBp4vmVor"
    rel="stylesheet"
  />,
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>,
]
`;

exports[`Missions component testing using snapshots Test Rockets component renders 1`] = `
Array [
  <link
    crossOrigin="anonymous"
    href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css"
    integrity="sha384-0evHe/X+R7YkIZDRvuzKMRqM+OrBnVFBL6DOitfPri4tjfHxaWutUpFmBp4vmVor"
    rel="stylesheet"
  />,
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>,
]
`;

exports[`Rockets component testing using snapshots Test Rockets component renders 1`] = `
Array [
  <link
    crossOrigin="anonymous"
    href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css"
    integrity="sha384-0evHe/X+R7YkIZDRvuzKMRqM+OrBnVFBL6DOitfPri4tjfHxaWutUpFmBp4vmVor"
    rel="stylesheet"
  />,
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>,
]
`;
